### PR TITLE
Port of VirtualConsole to LibVT, text mode

### PR DESCRIPTION
**Pull request by bugaevc**
cc @deoxxa @alimpfard 

There are now two different ways to get a text mode shell:

* `boot_mode=text init=/bin/Shell` — this will make the kernel spawn `/bin/Shell` as PID 4 instead of SystemServer. There'll be *nothing* but the Shell (and the kernel tasks) running. No environment variables set, so the shell will not do any fancy effects unless you set `TERM=xterm` ~and restart it~. No filesystems mounted but the root, so `/tmp` is on the root Ext2FS, `/proc` is empty, and devices in `/dev` are unopenable. (Run `mount -a` manually to get to a more standard environment.) No sockets in `/tmp/portal` are present, so you won't be able to, say, do hostname lookup, unless you start LookupServer manually the way SystemServer does it. This is as spartan as it gets.

* just `boot_mode=text` — recognized by the SystemServer; it will disable all the graphical services (such as the WindowServer and its many clients), and instead spawn a pair of comfy shells for you on tty0 and tty1. In this configuration, you do not run as PID 4 (SystemServer does), you do get `TERM=xterm`, all the filesystems are mounted, and a few background services that make sense for a text mode are enabled. This is for your regular usage in text mode.

And as always, you can customize what gets spawned on startup (and how) by tweaking `/etc/SystemServer.ini`. For example, if you're hacking on ~PaintBrush~ PixelPaint, why not replace the default Terminal with ~PaintBrush~ PixelPaint to have SystemServer start it up for you? Or if you want to provide remote access to a Serenity box, why not start TelnetServer when booting in text mode.

### CODE REVIEW

**Comment by alimpfard** [comment]
> unless you set TERM=xterm and restart it.

The line editor should pick it up on the following line after `export TERM=xterm`...

File: DevTools/HackStudio/CursorTool.cpp [hunk 1]
```diff
@@ -147,6 +147,8 @@ void CursorTool::on_keydown(GUI::KeyEvent& event)
         case Key_Right:
             move_selected_widgets_by(m_editor.form_widget().grid_size(), 0);
             break;
+        default:
+            break;
```
Comment (alimpfard) [inline DevTools/HackStudio/CursorTool.cpp:151]: I don't get your persistence on this, its existence makes no difference, and this exact behaviour is implied by its nonexistence.

Not nitpicking, just wondering out loud

File: Services/SystemServer/Service.cpp [hunk 2]
```diff
@@ -343,6 +348,16 @@ void Service::save_to(JsonObject& json)
...
+
+    JsonArray boot_modes;
+    for (String& mode : m_boot_modes)
+        boot_modes.append(mode);
+    json.set("boot_modes", boot_modes);
+
+    JsonArray environment;
+    for (String& env : m_environment)
+        boot_modes.append(env);
+    json.set("environment", environment);
```
Comment (alimpfard) [inline Services/SystemServer/Service.cpp:360]: :thinking: why is this stuff commented out?

Comment (alimpfard) [reply Services/SystemServer/Service.cpp:360]: Does passing a JsonArray crash the inspector?

**Comment by bugaevc** [comment]
> The line editor should pick it up on the following line after export TERM=xterm...

Oh, good. I think it didn't for me, but I only tested it once.

File: DevTools/HackStudio/CursorTool.cpp (hunk 1 above)
Comment (bugaevc) [reply DevTools/HackStudio/CursorTool.cpp:151]: 😀 it's not me, blame GCC

File: Services/SystemServer/Service.cpp (hunk 2 above)
Comment (bugaevc) [reply Services/SystemServer/Service.cpp:360]: An array, I think. At least it did when I first wrote that code.

File: Kernel/init.cpp [hunk 3]
```diff
@@ -302,28 +303,16 @@ void init_stage2()
...
-            hang();
-        }
-        thread->set_priority(THREAD_PRIORITY_HIGH);
+    // FIXME: It would be nicer to set the mode from userspace.
+    tty0->set_graphical(!text_mode);
+    Thread* thread = nullptr;
+    auto userspace_init = kernel_command_line().lookup("init").value_or("/bin/SystemServer");
+    Process::create_user_process(thread, userspace_init, (uid_t)0, (gid_t)0, (pid_t)0, error, {}, {}, tty0);
+    if (error != 0) {
+        klog() << "init_stage2: error spawning SystemServer: " << error;
```
Comment (bugaevc) [inline Kernel/init.cpp:312]: Should update the error message here I guess.

File: Kernel/init.cpp [hunk 4]
```diff
@@ -302,28 +303,16 @@ void init_stage2()
...
-        tty0->set_graphical(true);
-        Thread* thread = nullptr;
-        Process::create_user_process(thread, "/bin/SystemServer", (uid_t)0, (gid_t)0, (pid_t)0, error, {}, {}, tty0);
-        if (error != 0) {
-            klog() << "init_stage2: error spawning SystemServer: " << error;
-            hang();
-        }
-        thread->set_priority(THREAD_PRIORITY_HIGH);
+    // FIXME: It would be nicer to set the mode from userspace.
+    tty0->set_graphical(!text_mode);
```
Comment (bugaevc) [inline Kernel/init.cpp:307]: By that, I mean some `ioctl` on the tty

File: Services/SystemServer/Service.cpp (hunk 2 above)
Comment (alimpfard) [reply Services/SystemServer/Service.cpp:360]: I know that an empty array doesn't crash; let me check with an array that has something.

Comment (alimpfard) [reply Services/SystemServer/Service.cpp:360]: It seems to be okay with showing arrays (`[{"pid":25,"command":"te","running_time":5157,"pgid":25,"is_running_in_background":true}]` at least)

Comment (bugaevc) [reply Services/SystemServer/Service.cpp:360]: Why don't you uncomment the code here and try that.

Comment (alimpfard) [reply Services/SystemServer/Service.cpp:360]: Fiiine, I'll pull the PR

File: DevTools/HackStudio/CursorTool.cpp (hunk 1 above)
Comment (alimpfard) [reply DevTools/HackStudio/CursorTool.cpp:151]: Oh there are more enum values, I see.
\*blames GCC\*

File: Services/SystemServer/Service.cpp (hunk 2 above)
Comment (alimpfard) [reply Services/SystemServer/Service.cpp:360]: Seems OK to me.

**Comment by awesomekling** [comment]
This is super awesome :)

**Pull request merged**